# Data from NOAA and climod2
* https://www.ncdc.noaa.gov/cdo-web/
* http://climod2.nrcc.cornell.edu/
* Note:
    * T: Trace amount- Less than 0.01" precipitation; less than 0.1" snowfall; less than 1" snow depth.
    * S: Subsequent- Indicates the observation is missing, but is included in a subsequent value.
    * A: Accumulated- Indicates an accumulated value which includes the current day and any immediately preceeding missing days (starting with a day flagged by "S").
    * M: Missing.

___

# Importing Libraries and Data

In [1]:
import pandas as pd
import numpy as np
import matplotlib as plt
from matplotlib import pyplot as plt
%matplotlib inline
import seaborn as sns
import os
sns.set()

# Import NOAA data and grab useful columns

In [2]:
df = pd.read_csv('Bradley_1_8_1942_2_8_2023.csv', skipinitialspace = True)

In [3]:
df = df.drop(['STATION'], axis = 1)

In [4]:
df = df.drop(['NAME'], axis = 1)

In [5]:
df['Year'] = pd.DatetimeIndex(df['DATE']).year
df['Month'] = pd.DatetimeIndex(df['DATE']).month
df['Day'] = pd.DatetimeIndex(df['DATE']).day

In [6]:
df = df.sort_values(by=['DATE'])

In [7]:
for col in df.columns:
    print(col,'percent NaN:', df[col].isna().sum()/len(df))

DATE percent NaN: 0.0
ACMH percent NaN: 0.5988958824149935
ACSH percent NaN: 0.5987200675129224
AWND percent NaN: 0.4979078026653539
FMTM percent NaN: 0.6431660747564963
FRGT percent NaN: 0.9998593480783431
GAHT percent NaN: 0.9998945110587574
PGTM percent NaN: 0.5575090544674567
PRCP percent NaN: 0.0483842610499666
PSUN percent NaN: 0.7560040789057281
SNOW percent NaN: 0.12792292274693204
SNWD percent NaN: 0.30440592144590173
TAVG percent NaN: 0.7311086887724604
TMAX percent NaN: 0.04901719469742255
TMIN percent NaN: 0.04898203171700834
TSUN percent NaN: 0.5607088856851506
WDF1 percent NaN: 0.8460916347269595
WDF2 percent NaN: 0.6551566510777453
WDF5 percent NaN: 0.655754421744787
WDFG percent NaN: 0.7296670065754773
WDFM percent NaN: 0.7532965294138331
WESD percent NaN: 0.9402229332958262
WSF1 percent NaN: 0.846021308766131
WSF2 percent NaN: 0.6551566510777453
WSF5 percent NaN: 0.655754421744787
WSFG percent NaN: 0.729456028692992
WSFM percent NaN: 0.7532965294138331
WT01 percent NaN

# Drop
* ['ACMH ','ACSH ','AWND','FMTM','FRGT','GAHT','PGTM','PSUN','TSUN','WDF1','WDF2','WDF5','WDFG', 'WDFM','WESD','WSF1','WSF2','WSF5','WDFG','WDFM','WESD','WSF1','WSF2','WSF5','WSFG','WSFM', 'WT01','WT02','WT03','WT04','WT05','WT06','WT07','WT08','WT09', 'WT11','WT12','WT13','WT14','WT15','WT16','WT17','WT18','WT19', 'WT21','WT22','WV01','WV03']

# Keep
* DATE,ACMH,ACSH,AWND,FMTM,PGTM,PRCP,SNOW,SNWD,TAVG,TMAX,TMIN,TSUN,WDF1,WDF2,WDF5,WDFG,WDFM,
WESD,WSF1,WSF2,WSF5,WSFG,WSFM,WT01,WT08,WT16,Year,Month,Da

In [8]:
df = df.drop(['ACMH','ACSH','AWND','FMTM','FRGT','GAHT','PGTM','PSUN','TSUN','WDF1',
              'WDF2','WDF5','WDFG', 'WDFM','WESD','WSF1','WSF2','WSF5','WDFG','WDFM',
              'WESD','WSF1','WSF2','WSF5','WSFG','WSFM', 'WT01','WT02','WT03','WT04',
              'WT05','WT06','WT07','WT08','WT09', 'WT11','WT12','WT13','WT14','WT15',
              'WT16','WT17','WT18','WT19', 'WT21','WT22','WV01','WV03'], axis = 1).copy()

In [9]:
df.shape

(28439, 10)

In [10]:
for col in df.columns:
    print(col,'percent NaN:', df[col].isna().sum()/len(df))

DATE percent NaN: 0.0
PRCP percent NaN: 0.0483842610499666
SNOW percent NaN: 0.12792292274693204
SNWD percent NaN: 0.30440592144590173
TAVG percent NaN: 0.7311086887724604
TMAX percent NaN: 0.04901719469742255
TMIN percent NaN: 0.04898203171700834
Year percent NaN: 0.0
Month percent NaN: 0.0
Day percent NaN: 0.0


In [11]:
# lowercase all columns
df.columns = [x.lower() for x in df.columns]

# End Getting NOAA Data

___

# Import Cornell data and get useful columns

In [12]:
df_cornell = pd.read_csv('Bradley_Cornell_1_1_1940_2_13_2023.txt', sep=',', header=0)

In [13]:
df_cornell.columns = [i.replace(' ','') for i in df_cornell.columns]

In [14]:
df_cornell = df_cornell.sort_values(by=['Date'])

In [15]:
df_cornell.iloc[0,]

Date                       1940-01-01
MaxTemperature                      M
MaxTemperatureNormal             36.6
MaxTemperatureDeparture             M
MinTemperature                      M
MinTemperatureNormal             20.8
MinTemperatureDeparture             M
AvgTemperature                      M
AvgTemperatureNormal             28.7
AvgTemperatureDeparture             M
AtObsTemperature                    M
Precipitation                       M
PrecipitationNormal              0.11
PrecipitationDeparture              M
Snowfall                            M
SnowfallNormal                    0.4
SnowfallDeparture                   M
SnowDepth                           M
HDD                                 M
HDDNormal                        36.3
HDDDeparture                        M
CDD                                 M
CDDNormal                         0.0
CDDDeparture                        M
GDD                                 M
Name: 0, dtype: object

In [16]:
df_cornell['Year'] = pd.DatetimeIndex(df_cornell['Date']).year
df_cornell['Month'] = pd.DatetimeIndex(df_cornell['Date']).month
df_cornell['Day'] = pd.DatetimeIndex(df_cornell['Date']).day

In [17]:
for col in df_cornell.columns:
    print(col,'percent NaN:', df_cornell[col].isna().sum()/len(df_cornell))

Date percent NaN: 0.0
MaxTemperature percent NaN: 0.0
MaxTemperatureNormal percent NaN: 0.0
MaxTemperatureDeparture percent NaN: 0.0
MinTemperature percent NaN: 0.0
MinTemperatureNormal percent NaN: 0.0
MinTemperatureDeparture percent NaN: 0.0
AvgTemperature percent NaN: 0.0
AvgTemperatureNormal percent NaN: 0.0
AvgTemperatureDeparture percent NaN: 0.0
AtObsTemperature percent NaN: 0.0
Precipitation percent NaN: 0.0
PrecipitationNormal percent NaN: 0.0
PrecipitationDeparture percent NaN: 0.0
Snowfall percent NaN: 0.0
SnowfallNormal percent NaN: 0.0
SnowfallDeparture percent NaN: 0.0
SnowDepth percent NaN: 0.0
HDD percent NaN: 0.0
HDDNormal percent NaN: 0.0
HDDDeparture percent NaN: 0.0
CDD percent NaN: 0.0
CDDNormal percent NaN: 0.0
CDDDeparture percent NaN: 0.0
GDD percent NaN: 0.0
Year percent NaN: 0.0
Month percent NaN: 0.0
Day percent NaN: 0.0


In [18]:
df_cornell.head(1)

Date MaxTemperature  MaxTemperatureNormal MaxTemperatureDeparture  \
0  1940-01-01              M                  36.6                       M   

  MinTemperature  MinTemperatureNormal MinTemperatureDeparture AvgTemperature  \
0              M                  20.8                       M              M   

   AvgTemperatureNormal AvgTemperatureDeparture  ... HDD HDDNormal  \
0                  28.7                       M  ...   M      36.3   

  HDDDeparture CDD CDDNormal CDDDeparture GDD  Year Month  Day  
0            M   M       0.0            M   M  1940     1    1  

[1 rows x 28 columns]

In [19]:
df_cornell.iloc[0,1]

' M'

In [20]:
for col in df_cornell.columns:
    print(col, len(df_cornell[df_cornell[col].isin([' M'])])/len(df_cornell))

Date 0.0
MaxTemperature 0.10912384716732543
MaxTemperatureNormal 0.0
MaxTemperatureDeparture 0.10912384716732543
MinTemperature 0.10909090909090909
MinTemperatureNormal 0.0
MinTemperatureDeparture 0.10909090909090909
AvgTemperature 0.10922266139657444
AvgTemperatureNormal 0.0
AvgTemperatureDeparture 0.10922266139657444
AtObsTemperature 1.0
Precipitation 0.10843214756258235
PrecipitationNormal 0.000691699604743083
PrecipitationDeparture 0.10902503293807642
Snowfall 0.18158761528326745
SnowfallNormal 0.000691699604743083
SnowfallDeparture 0.18214756258234518
SnowDepth 0.2326086956521739
HDD 0.10922266139657444
HDDNormal 0.0
HDDDeparture 0.10922266139657444
CDD 0.10922266139657444
CDDNormal 0.0
CDDDeparture 0.10922266139657444
GDD 0.10922266139657444
Year 0.0
Month 0.0
Day 0.0


In [21]:
for col in df_cornell.columns:
    print(col, len(df_cornell[df_cornell[col].isin([' T'])])/len(df_cornell))

Date 0.0
MaxTemperature 0.0
MaxTemperatureNormal 0.0
MaxTemperatureDeparture 0.0
MinTemperature 0.0
MinTemperatureNormal 0.0
MinTemperatureDeparture 0.0
AvgTemperature 0.0
AvgTemperatureNormal 0.0
AvgTemperatureDeparture 0.0
AtObsTemperature 0.0
Precipitation 0.12262845849802372
PrecipitationNormal 0.0
PrecipitationDeparture 0.0
Snowfall 0.056357048748353095
SnowfallNormal 0.0
SnowfallDeparture 0.0
SnowDepth 0.03593544137022398
HDD 0.0
HDDNormal 0.0
HDDDeparture 0.0
CDD 0.0
CDDNormal 0.0
CDDDeparture 0.0
GDD 0.0
Year 0.0
Month 0.0
Day 0.0


In [22]:
for col in df_cornell.columns:
    print(col, len(df_cornell[df_cornell[col].isin([' S'])])/len(df_cornell))

Date 0.0
MaxTemperature 0.0
MaxTemperatureNormal 0.0
MaxTemperatureDeparture 0.0
MinTemperature 0.0
MinTemperatureNormal 0.0
MinTemperatureDeparture 0.0
AvgTemperature 0.0
AvgTemperatureNormal 0.0
AvgTemperatureDeparture 0.0
AtObsTemperature 0.0
Precipitation 0.0
PrecipitationNormal 0.0
PrecipitationDeparture 0.0
Snowfall 0.0
SnowfallNormal 0.0
SnowfallDeparture 0.0
SnowDepth 0.0
HDD 0.0
HDDNormal 0.0
HDDDeparture 0.0
CDD 0.0
CDDNormal 0.0
CDDDeparture 0.0
GDD 0.0
Year 0.0
Month 0.0
Day 0.0


In [23]:
for col in df_cornell.columns:
    print(col, len(df_cornell[df_cornell[col].isin([' A'])])/len(df_cornell))

Date 0.0
MaxTemperature 0.0
MaxTemperatureNormal 0.0
MaxTemperatureDeparture 0.0
MinTemperature 0.0
MinTemperatureNormal 0.0
MinTemperatureDeparture 0.0
AvgTemperature 0.0
AvgTemperatureNormal 0.0
AvgTemperatureDeparture 0.0
AtObsTemperature 0.0
Precipitation 0.0
PrecipitationNormal 0.0
PrecipitationDeparture 0.0
Snowfall 0.0
SnowfallNormal 0.0
SnowfallDeparture 0.0
SnowDepth 0.0
HDD 0.0
HDDNormal 0.0
HDDDeparture 0.0
CDD 0.0
CDDNormal 0.0
CDDDeparture 0.0
GDD 0.0
Year 0.0
Month 0.0
Day 0.0


# Keep:
* Date, MaxTemperature, MinTemperature, AvgTemperature, AtObsTemperature, Precipitation, Snowfall, SnowDepth,
* Year, Month, Day
# Drop: 
* MaxTemperatureNormal, MaxTemperatureDeparture, MinTemperatureNormal, MinTemperatureDeparture, AvgTemperatureNormal
* AvgTemperatureDeparture, PrecipitationNormal, PrecipitationDeparture, SnowfallNormal, SnowfallDeparture
* HDD, HDDNormal, HDDDeparture, CDD, CDDNormal, CDDDeparture, GDD

In [24]:
df_cornell.shape

(30360, 28)

In [25]:
df_cornell.columns

Index(['Date', 'MaxTemperature', 'MaxTemperatureNormal',
       'MaxTemperatureDeparture', 'MinTemperature', 'MinTemperatureNormal',
       'MinTemperatureDeparture', 'AvgTemperature', 'AvgTemperatureNormal',
       'AvgTemperatureDeparture', 'AtObsTemperature', 'Precipitation',
       'PrecipitationNormal', 'PrecipitationDeparture', 'Snowfall',
       'SnowfallNormal', 'SnowfallDeparture', 'SnowDepth', 'HDD', 'HDDNormal',
       'HDDDeparture', 'CDD', 'CDDNormal', 'CDDDeparture', 'GDD', 'Year',
       'Month', 'Day'],
      dtype='object')

In [26]:
df_cornell = df_cornell.drop(['MaxTemperatureNormal','MaxTemperatureDeparture','MinTemperatureNormal',
                             'MinTemperatureDeparture','AvgTemperatureNormal','AvgTemperatureDeparture',
                             'PrecipitationNormal', 'PrecipitationDeparture','SnowfallNormal', 'SnowfallDeparture',
                             'HDD', 'HDDNormal','HDDDeparture', 'CDD', 'CDDNormal', 'CDDDeparture', 'GDD'], axis = 1).copy()

In [27]:
df_cornell.shape

(30360, 11)

In [28]:
# lowercase all columns
df_cornell.columns = [x.lower() for x in df_cornell.columns]

# End Getting Cornell Data

___

# Check date range and make sure the same

In [29]:
df_cornell.iloc[0,0]

'1940-01-01'

In [30]:
type(df_cornell.iloc[0,0])

str

In [31]:
df.iloc[0,0]

'1942-01-08'

In [32]:
type(df.iloc[0,0])

str

In [33]:
print(df.date.min())
print(df.date.max())
print(df_cornell.date.min())
print(df_cornell.date.max())

1942-01-08
2023-02-08
1940-01-01
2023-02-13


In [34]:
df = df[(df['date'] >= '1943-01-01') & (df['date'] < '2023-01-01')].copy()

In [35]:
df_cornell = df_cornell[(df_cornell['date'] >= '1943-01-01') & (df_cornell['date'] < '2023-01-01')].copy()

In [36]:
df.head(2)

date  prcp  snow  snwd  tavg  tmax  tmin  year  month  day
356  1943-01-01   NaN   NaN   NaN  30.0   NaN   NaN  1943      1    1
357  1943-01-02   NaN   NaN   NaN  25.0   NaN   NaN  1943      1    2

In [37]:
df_cornell.head(2)

date maxtemperature mintemperature avgtemperature  \
1096  1943-01-01              M              M              M   
1097  1943-01-02              M              M              M   

     atobstemperature precipitation snowfall snowdepth  year  month  day  
1096                M             M        M         M  1943      1    1  
1097                M             M        M         M  1943      1    2

In [38]:
print(df.shape)
print(df_cornell.shape)

(28044, 10)
(29220, 11)


### NOAA is missing data or Cornell has duplicate dates

In [39]:
for noaa, cornell in zip(df.date, df_cornell.date):
    if noaa != cornell:
        print(noaa, cornell)
        break

1949-01-01 1945-10-15


In [40]:
print(len(set(df.date)))
print(len(set(df_cornell.date)))

28044
29220


In [41]:
date_diff = set(df_cornell.date).difference(set(df.date))
print(len(date_diff))
print(len(set(df_cornell.date))-len(set(df.date)))

1176
1176


In [42]:
df_cornell[(df_cornell['date']>='1945-10-12') & (df_cornell['date']<'1945-10-22')]

date maxtemperature mintemperature avgtemperature  \
2111  1945-10-12              M              M              M   
2112  1945-10-13              M              M              M   
2113  1945-10-14              M              M              M   
2114  1945-10-15              M              M              M   
2115  1945-10-16              M              M              M   
2116  1945-10-17              M              M              M   
2117  1945-10-18              M              M              M   
2118  1945-10-19              M              M              M   
2119  1945-10-20              M              M              M   
2120  1945-10-21              M              M              M   

     atobstemperature precipitation snowfall snowdepth  year  month  day  
2111                M             M        M         M  1945     10   12  
2112                M             M        M         M  1945     10   13  
2113                M             M        M         M  1945     10   14  
2114                M             M        M         M  1945     10   15  
2115                M             M        M         M  1945     10   16  
2116                M             M        M         M  1945     10   17  
2117                M             M        M         M  1945     10   18  
2118                M             M        M         M  1945     10   19  
2119                M             M        M         M  1945     10   20  
2120                M             M        M         M  1945     10   21

In [43]:
df[(df['date']>='1945-10-12') & (df['date']<'1947-10-30')]

date  prcp  snow  snwd  tavg  tmax  tmin  year  month  day
1371  1945-10-12   NaN   NaN   NaN  49.0   NaN   NaN  1945     10   12
1372  1945-10-13   NaN   NaN   NaN  53.0   NaN   NaN  1945     10   13
1373  1945-10-14   NaN   NaN   NaN  48.0   NaN   NaN  1945     10   14

In [44]:
# looks like some years are missing from NOAA
diff_years = set(df_cornell.year).difference(set(df.year))
print(diff_years)

{1946, 1947, 1948}


In [45]:
cor_46_47_48 = df_cornell[(df_cornell['date'] >= '1946-01-01') & (df_cornell['date'] < '1949-01-01')].copy()

In [46]:
cor_46_47_48

date maxtemperature mintemperature avgtemperature  \
2192  1946-01-01              M              M              M   
2193  1946-01-02              M              M              M   
2194  1946-01-03              M              M              M   
2195  1946-01-04              M              M              M   
2196  1946-01-05              M              M              M   
...          ...            ...            ...            ...   
3283  1948-12-27              M              M              M   
3284  1948-12-28              M              M              M   
3285  1948-12-29              M              M              M   
3286  1948-12-30              M              M              M   
3287  1948-12-31              M              M              M   

     atobstemperature precipitation snowfall snowdepth  year  month  day  
2192                M             M        M         M  1946      1    1  
2193                M             M        M         M  1946      1    2  
2194                M             M        M         M  1946      1    3  
2195                M             M        M         M  1946      1    4  
2196                M             M        M         M  1946      1    5  
...               ...           ...      ...       ...   ...    ...  ...  
3283                M             M        M         M  1948     12   27  
3284                M             M        M         M  1948     12   28  
3285                M             M        M         M  1948     12   29  
3286                M             M        M         M  1948     12   30  
3287                M             M        M         M  1948     12   31  

[1096 rows x 11 columns]

# Looks like there is a lot of missing value so let's check when actual data comes in

# 1940s

In [47]:
df_40s_dates = ['1940-01-01','1941-01-01','1942-01-01','1943-01-01','1944-01-01','1945-01-01',
               '1946-01-01','1947-01-01','1948-01-01','1949-01-01','1950-01-01']

In [48]:
for start, end in zip(df_40s_dates[0:-1],df_40s_dates[1:]):
    df_40s = df[(df['date'] >= start) & (df['date'] < end)].copy()
    print(start)
    for col in df_40s.columns:
        try:
            if df_40s[col].isna().sum()/len(df_40s) > 0:
                print(col,'percent na: ',df_40s[col].isna().sum()/len(df_40s))
        except:
            continue

1940-01-01
1941-01-01
1942-01-01
1943-01-01
prcp percent na:  1.0
snow percent na:  1.0
snwd percent na:  1.0
tmax percent na:  1.0
tmin percent na:  1.0
1944-01-01
prcp percent na:  1.0
snow percent na:  1.0
snwd percent na:  1.0
tmax percent na:  1.0
tmin percent na:  1.0
1945-01-01
prcp percent na:  1.0
snow percent na:  1.0
snwd percent na:  1.0
tmax percent na:  1.0
tmin percent na:  1.0
1946-01-01
1947-01-01
1948-01-01
1949-01-01
tavg percent na:  1.0
tmax percent na:  0.0027397260273972603


C:\Users\mpica\AppData\Local\Temp/ipykernel_13764/387063338.py:6: RuntimeWarning: invalid value encountered in longlong_scalars
  if df_40s[col].isna().sum()/len(df_40s) > 0:


# NOAA
* Dropped Data for 1940 to 1942
* No data for 1946, 1947, 1948,
* 1943:
    * prcp percent na:  1.0
    * snow percent na:  1.0
    * snwd percent na:  1.0
    * tmax percent na:  1.0
    * tmin percent na:  1.0
* 1944
    * prcp percent na:  1.0
    * snow percent na:  1.0
    * snwd percent na:  1.0
    * tmax percent na:  1.0
    * tmin percent na:  1.0
* 1945
    * prcp percent na:  1.0
    * snow percent na:  1.0
    * snwd percent na:  1.0
    * tmax percent na:  1.0
    * tmin percent na:  1.0
* 1949
    * tavg percent na:  1.0
    * tmax percent na:  0.0027397260273972603

In [49]:
for start, end in zip(df_40s_dates[0:-1],df_40s_dates[1:]):
    df_40s = df_cornell[(df_cornell['date'] >= start) & (df_cornell['date'] < end)].copy()
    print(start)
    for col in df_40s.columns:
        try:
            if df_40s[col].isna().sum()/len(df_40s) > 0:
                print(col,'percent na: ',df_40s[col].isna().sum()/len(df_40s))
        except:
            continue

1940-01-01
1941-01-01
1942-01-01
1943-01-01
1944-01-01
1945-01-01
1946-01-01


C:\Users\mpica\AppData\Local\Temp/ipykernel_13764/324951126.py:6: RuntimeWarning: invalid value encountered in longlong_scalars
  if df_40s[col].isna().sum()/len(df_40s) > 0:


1947-01-01
1948-01-01
1949-01-01


In [64]:
bad_chars = [' S',' M',' T',' A','S','M','T','A']
for start, end in zip(df_40s_dates[0:-1],df_40s_dates[1:]):
    print(start)
    df_40s_cornell = df_cornell[(df_cornell['date'] >= start) & (df_cornell['date'] < end)].copy()
    for col in df_40s_cornell.columns:
        count = 0
        for val in df_40s_cornell[col].to_numpy():
            try:
                val = val.upper()
            except:
                continue
            if any(x in bad_chars for x in val):
                count += 1
        if count > 0:
            print(col,'percent bad:', round(count/len(df_40s_cornell)*100,4))
    del df_40s_cornell

1940-01-01
1941-01-01
1942-01-01
1943-01-01
maxtemperature percent bad: 100.0
mintemperature percent bad: 100.0
avgtemperature percent bad: 100.0
atobstemperature percent bad: 100.0
precipitation percent bad: 100.0
snowfall percent bad: 100.0
snowdepth percent bad: 100.0
1944-01-01
maxtemperature percent bad: 100.0
mintemperature percent bad: 100.0
avgtemperature percent bad: 100.0
atobstemperature percent bad: 100.0
precipitation percent bad: 100.0
snowfall percent bad: 100.0
snowdepth percent bad: 100.0
1945-01-01
maxtemperature percent bad: 100.0
mintemperature percent bad: 100.0
avgtemperature percent bad: 100.0
atobstemperature percent bad: 100.0
precipitation percent bad: 100.0
snowfall percent bad: 100.0
snowdepth percent bad: 100.0
1946-01-01
maxtemperature percent bad: 100.0
mintemperature percent bad: 100.0
avgtemperature percent bad: 100.0
atobstemperature percent bad: 100.0
precipitation percent bad: 100.0
snowfall percent bad: 100.0
snowdepth percent bad: 100.0
1947-01-01


# Cornell
* Dropped Data for 1940 to 1942
* 1943
    * maxtemperature percent bad: 100.0
    * mintemperature percent bad: 100.0
    * avgtemperature percent bad: 100.0
    * atobstemperature percent bad: 100.0
    * precipitation percent bad: 100.0
    * snowfall percent bad: 100.0
    * snowdepth percent bad: 100.0
* 1944
    * maxtemperature percent bad: 100.0
    * mintemperature percent bad: 100.0
    * avgtemperature percent bad: 100.0
    * atobstemperature percent bad: 100.0
    * precipitation percent bad: 100.0
    * snowfall percent bad: 100.0
    * snowdepth percent bad: 100.0
* 1945
    * maxtemperature percent bad: 100.0
    * mintemperature percent bad: 100.0
    * avgtemperature percent bad: 100.0
    * atobstemperature percent bad: 100.0
    * precipitation percent bad: 100.0
    * snowfall percent bad: 100.0
    * snowdepth percent bad: 100.0
* 1946
    * maxtemperature percent bad: 100.0
    * mintemperature percent bad: 100.0
    * avgtemperature percent bad: 100.0
    * atobstemperature percent bad: 100.0
    * precipitation percent bad: 100.0
    * snowfall percent bad: 100.0
    * snowdepth percent bad: 100.0
* 1947
    * maxtemperature percent bad: 100.0
    * mintemperature percent bad: 100.0
    * avgtemperature percent bad: 100.0
    * atobstemperature percent bad: 100.0
    * precipitation percent bad: 100.0
    * snowfall percent bad: 100.0
    * snowdepth percent bad: 100.0
* 1948
    * maxtemperature percent bad: 100.0
    * mintemperature percent bad: 100.0
    * avgtemperature percent bad: 100.0
    * atobstemperature percent bad: 100.0
    * precipitation percent bad: 100.0
    * snowfall percent bad: 100.0
    * snowdepth percent bad: 100.0
* 1949
    * maxtemperature percent bad: 0.8219
    * mintemperature percent bad: 0.5479
    * avgtemperature percent bad: 1.3699
    * atobstemperature percent bad: 100.0
    * precipitation percent bad: 16.4384
    * snowfall percent bad: 5.7534
    * snowdepth percent bad: 2.4658

In [54]:
df.head(1)

date  prcp  snow  snwd  tavg  tmax  tmin  year  month  day
356  1943-01-01   NaN   NaN   NaN  30.0   NaN   NaN  1943      1    1

In [55]:
df_cornell.head(1)

date maxtemperature mintemperature avgtemperature  \
1096  1943-01-01              M              M              M   

     atobstemperature precipitation snowfall snowdepth  year  month  day  
1096                M             M        M         M  1943      1    1

In [ ]:
* NOAA
* Dropped Data for 1940 to 1942
* No data for 1946, 1947, 1948,
* NOAA (tavg)
    * 1943:
        * prcp percent na:  1.0
        * snow percent na:  1.0
        * snwd percent na:  1.0
        * tmax percent na:  1.0
        * tmin percent na:  1.0
* cornell (all bad)
    * 1943
        * maxtemperature percent bad: 100.0
        * mintemperature percent bad: 100.0
        * avgtemperature percent bad: 100.0
        * atobstemperature percent bad: 100.0
        * precipitation percent bad: 100.0
        * snowfall percent bad: 100.0
        * snowdepth percent bad: 100.0
* NOAA (tavg)
    * 1944
        * prcp percent na:  1.0
        * snow percent na:  1.0
        * snwd percent na:  1.0
        * tmax percent na:  1.0
        * tmin percent na:  1.0
* Cornell (all bad)
    * 1944
        * maxtemperature percent bad: 100.0
        * mintemperature percent bad: 100.0
        * avgtemperature percent bad: 100.0
        * atobstemperature percent bad: 100.0
        * precipitation percent bad: 100.0
        * snowfall percent bad: 100.0
        * snowdepth percent bad: 100.0
* NOAA (tavg)
    * 1945
        * prcp percent na:  1.0
        * snow percent na:  1.0
        * snwd percent na:  1.0
        * tmax percent na:  1.0
        * tmin percent na:  1.0
* cornell (all bad)
    * 1945
        * maxtemperature percent bad: 100.0
        * mintemperature percent bad: 100.0
        * avgtemperature percent bad: 100.0
        * atobstemperature percent bad: 100.0
        * precipitation percent bad: 100.0
        * snowfall percent bad: 100.0
        * snowdepth percent bad: 100.0
* NOAA
    * 1946: NO DATA
* cornell (all bad)
    * 1946
        * maxtemperature percent bad: 100.0
        * mintemperature percent bad: 100.0
        * avgtemperature percent bad: 100.0
        * atobstemperature percent bad: 100.0
        * precipitation percent bad: 100.0
        * snowfall percent bad: 100.0
        * snowdepth percent bad: 100.0
* NOAA
    * 1947: NO DATA
* Cornell (all bad)
    * 1947
        * maxtemperature percent bad: 100.0
        * mintemperature percent bad: 100.0
        * avgtemperature percent bad: 100.0
        * atobstemperature percent bad: 100.0
        * precipitation percent bad: 100.0
        * snowfall percent bad: 100.0
        * snowdepth percent bad: 100.0
* NOAA
    * 1948: NO DATA
* Cornell (all bad)
    * 1948
        * maxtemperature percent bad: 100.0
        * mintemperature percent bad: 100.0
        * avgtemperature percent bad: 100.0
        * atobstemperature percent bad: 100.0
        * precipitation percent bad: 100.0
        * snowfall percent bad: 100.0
        * snowdepth percent bad: 100.0
* NOAA (prcp,snow,snwd,tmin)
    * 1949
        * tavg percent na:  1.0
        * tmax percent na:  0.0027397260273972603
* cornell
    * 1949
        * maxtemperature percent bad: 0.8219
        * mintemperature percent bad: 0.5479
        * avgtemperature percent bad: 1.3699
        * atobstemperature percent bad: 100.0
        * precipitation percent bad: 16.4384
        * snowfall percent bad: 5.7534
        * snowdepth percent bad: 2.4658

In [78]:
df[(df['tavg'].isna()) & (df['date']>='1940-01-01') & (df['date']<'1950-01-01')].shape

(365, 10)

In [108]:
df[(df['tmax'].isna()) & (df['date']>='1949-01-01') & (df['date']<'1950-01-01')]

date  prcp  snow  snwd  tavg  tmax  tmin  year  month  day
1395  1949-01-22  0.27   2.0   1.0   NaN   NaN  29.0  1949      1   22

In [106]:
df[(df['tmin'].isna()) & (df['date']>='1949-01-01') & (df['date']<'1950-01-01')].shape

(0, 10)

In [107]:
df[(df['tavg'].isna()) & (df['date']>='1949-01-01') & (df['date']<'1950-01-01')].shape

(365, 10)

In [109]:
df_cornell[(df_cornell['avgtemperature'].isin([' M'])) & (df_cornell['date']>='1949-01-01') & (df_cornell['date']<'1950-01-01')]

date maxtemperature mintemperature avgtemperature  \
3309  1949-01-22              M             29              M   
3591  1949-10-31             64              M              M   
3592  1949-11-01              M             40              M   
3616  1949-11-25             48              M              M   
3617  1949-11-26              M             17              M   

     atobstemperature precipitation snowfall snowdepth  year  month  day  
3309                M          0.27      2.0         1  1949      1   22  
3591                M          0.50      0.0         0  1949     10   31  
3592                M          0.01      0.0         0  1949     11    1  
3616                M          0.35      0.0         0  1949     11   25  
3617                M          0.00      0.0         0  1949     11   26

## Best Data for 1940s

* 1949 ONLY      
    * prcp (precipitation): NOAA: No missing
    * snow (snowfall): NOAA: No missing
    * snwd (snowdepth): NOAA: No missing
    * tavg (avgtemperature): Need to see if they don't have the same day missing
        * They both have min temp but no max temp to get average for 1/22/1949
    * tmax (maxtemperature): Need to see if they don't have the same day missing
        * They are both missing data for max temp on 1/22/2949
    * tmin (mintemperature): NOAA: No missing
    * atobstemperature: cornell: atobstemperature percent bad: 100%

# 1950s

In [56]:
df_50s_dates = ['1950-01-01','1951-01-01','1952-01-01','1953-01-01','1954-01-01','1955-01-01',
               '1956-01-01','1957-01-01','1958-01-01','1959-01-01','1960-01-01']

In [57]:
# NOAA
for start, end in zip(df_50s_dates[0:-1],df_50s_dates[1:]):
    df_50s = df[(df['date'] >= start) & (df['date'] < end)].copy()
    print(start)
    for col in df_50s.columns:
        try:
            if df_50s[col].isna().sum()/len(df_50s) > 0:
                print(col,'percent na: ',df_50s[col].isna().sum()/len(df_50s))
        except:
            continue

1950-01-01
snwd percent na:  0.005479452054794521
tavg percent na:  1.0
1951-01-01
tavg percent na:  1.0
1952-01-01
tavg percent na:  1.0
1953-01-01
snwd percent na:  0.0136986301369863
tavg percent na:  1.0
tmax percent na:  0.005479452054794521
tmin percent na:  0.005479452054794521
1954-01-01
tavg percent na:  1.0
1955-01-01
tavg percent na:  1.0
1956-01-01
snow percent na:  0.00273224043715847
snwd percent na:  0.00273224043715847
tavg percent na:  1.0
1957-01-01
snwd percent na:  0.005494505494505495
tavg percent na:  1.0
1958-01-01
tavg percent na:  1.0
1959-01-01
snwd percent na:  0.0027397260273972603
tavg percent na:  1.0


In [58]:
for start, end in zip(df_50s_dates[0:-1],df_50s_dates[1:]):
    df_50s = df_cornell[(df_cornell['date'] >= start) & (df_cornell['date'] < end)].copy()
    print(start)
    for col in df_50s.columns:
        try:
            if df_50s[col].isna().sum()/len(df_50s) > 0:
                print(col,'percent na: ',df_50s[col].isna().sum()/len(df_40s))
        except:
            continue

1950-01-01
1951-01-01
1952-01-01
1953-01-01
1954-01-01
1955-01-01
1956-01-01
1957-01-01
1958-01-01
1959-01-01


In [61]:
bad_chars = [' S',' M',' T',' A','S','M','T','A']
for start, end in zip(df_50s_dates[0:-1],df_50s_dates[1:]):
    print(start)
    df_50s_cornell = df_cornell[(df_cornell['date'] >= start) & (df_cornell['date'] < end)].copy()
    for col in df_50s_cornell.columns:
        count = 0
        for val in df_50s_cornell[col].to_numpy():
            try:
                val = val.upper()
            except:
                continue
            if any(x in bad_chars for x in val):
                count += 1
        if count > 0:
            print(col,'percent bad:', round(count/len(df_50s_cornell)*100,4))
    del df_50s_cornell

1950-01-01
atobstemperature percent bad: 100.0
precipitation percent bad: 16.4384
snowfall percent bad: 10.6849
snowdepth percent bad: 2.1918
1951-01-01
maxtemperature percent bad: 0.274
mintemperature percent bad: 0.274
avgtemperature percent bad: 0.5479
atobstemperature percent bad: 100.0
precipitation percent bad: 18.0822
snowfall percent bad: 7.1233
snowdepth percent bad: 1.6438
1952-01-01
atobstemperature percent bad: 100.0
precipitation percent bad: 15.3005
snowfall percent bad: 7.6503
snowdepth percent bad: 4.0984
1953-01-01
maxtemperature percent bad: 0.5479
mintemperature percent bad: 0.5479
avgtemperature percent bad: 0.5479
atobstemperature percent bad: 100.0
precipitation percent bad: 13.6986
snowfall percent bad: 10.6849
snowdepth percent bad: 3.0137
1954-01-01
atobstemperature percent bad: 100.0
precipitation percent bad: 17.8082
snowfall percent bad: 6.8493
snowdepth percent bad: 6.0274
1955-01-01
atobstemperature percent bad: 100.0
precipitation percent bad: 12.3288
sno

In [60]:
df_cornell.shape

(29220, 11)

* 1950
    * prcp (precipitation): NOAA: No missing
    * snow (snowfall): NOAA: No missing
    * snwd (snowdepth): NOAA: 0.00548% missing Cornell: 2.2% bad
    * tavg (avgtemperature): Cornell: No missing
    * tmax (maxtemperature): Need to see if they don't have the same day missing
    * tmin (mintemperature): NOAA: No missing
    * atobstemperature: cornell: atobstemperature percent bad: 100%
* 1951
    * prcp (precipitation): NOAA: No missing
    * snow (snowfall): NOAA: No missing
    * snwd (snowdepth): NOAA: 0.00548% missing Cornell: 2.2% bad
    * tavg (avgtemperature): Cornell: 0.5479 missing
    * tmax (maxtemperature): Need to see if they don't have the same day missing
    * tmin (mintemperature): NOAA: No missing
    * atobstemperature: cornell: atobstemperature percent bad: 100%
* 1952
    * prcp (precipitation): NOAA: No missing
    * snow (snowfall): NOAA: No missing
    * snwd (snowdepth): NOAA: No missing
    * tavg (avgtemperature): Cornell: No missing
    * tmax (maxtemperature): Cornell: No missing
    * tmin (mintemperature): Cornell: No missing
    * atobstemperature: cornell: atobstemperature percent bad: 100%
* 1953
    * prcp (precipitation): NOAA: No missing
    * snow (snowfall): NOAA: No missing
    * snwd (snowdepth): NOAA: 0.0136% missing, Cornell: 3% missing
    * tavg (avgtemperature): Cornell: 0.5479% missing, NOAA: All missing
    * tmax (maxtemperature): Cornell: 0.5479% missing, NOAA: 0.0055% missing
    * tmin (mintemperature): Cornell: 0.5479% missing, NOAA: 0.0055% missing
    * atobstemperature: cornell: atobstemperature percent bad: 100%
* 1954
    * prcp (precipitation): NOAA: No missing
    * snow (snowfall): NOAA: No missing
    * snwd (snowdepth): NOAA: 0.0136% missing, Cornell: 3% missing
    * tavg (avgtemperature): Cornell: No missing
    * tmax (maxtemperature): Cornell: No missing, NOAA: No missing
    * tmin (mintemperature): Cornell: No missing, NOAA: No missing
    * atobstemperature: cornell: atobstemperature percent bad: 100%
* 1955
    * prcp (precipitation): NOAA: No missing
    * snow (snowfall): NOAA: No missing
    * snwd (snowdepth): NOAA: 0.0136% missing, Cornell: 3% missing
    * tavg (avgtemperature): Cornell: No missing
    * tmax (maxtemperature): Cornell: No missing, NOAA: No missing
    * tmin (mintemperature): Cornell: No missing, NOAA: No missing
    * atobstemperature: cornell: atobstemperature percent bad: 100%
* 1956
    * prcp (precipitation): NOAA: No missing
    * snow (snowfall): NOAA: 0.0027% missing, Cornell: 6.0109% missing
    * snwd (snowdepth): NOAA: 0.0027% missing, Cornell: 9.2896% missing
    * tavg (avgtemperature): Cornell: No missing
    * tmax (maxtemperature): Cornell: No missing, NOAA: No missing
    * tmin (mintemperature): Cornell: No missing, NOAA: No missing
    * atobstemperature: cornell: atobstemperature percent bad: 100%
* 1957
    * prcp (precipitation): NOAA: No missing
    * snow (snowfall): NOAA: No missing
    * snwd (snowdepth): NOAA: 5.7534% missing, Cornell: 0.005% missing
    * tavg (avgtemperature): Cornell: 0.274% missing, NOAA: All missing
        * Use NOAA max,min to get average for missing days if overlap doesn't work
    * tmax (maxtemperature): NOAA: No missing
    * tmin (mintemperature): NOAA: No missing
    * atobstemperature: cornell: atobstemperature percent bad: 100%
* 1958
    * prcp (precipitation): NOAA: No missing
    * snow (snowfall): NOAA: No missing
    * snwd (snowdepth): NOAA: No missing
    * tavg (avgtemperature): Cornell: No missing
    * tmax (maxtemperature): NOAA: No missing
    * tmin (mintemperature): NOAA: No missing
    * atobstemperature: cornell: atobstemperature percent bad: 100%
* 1959
    * prcp (precipitation): NOAA: No missing
    * snow (snowfall): NOAA: No missing
    * snwd (snowdepth): NOAA: 0.0027% missing, Cornell: 5.7534% missing
    * tavg (avgtemperature): Cornell: No missing
    * tmax (maxtemperature): NOAA: No missing
    * tmin (mintemperature): NOAA: No missing
    * atobstemperature: cornell: atobstemperature percent bad: 100%